modules/loader-utils/src/lib/request-utils/request-scheduler.js

```diff
@@ -44,59 +44,63 @@ export default class RequestScheduler {
   }
 
   // Called by an application that wants to issue a request, without having it deeply queued
-  // Parameter `callback` will be called when request "slots" open up,
+  // Parameter `getPriority` will be called when request "slots" open up,
   //    allowing the caller to update priority or cancel the request
   //    Highest priority executes first, priority < 0 cancels the request
-  // Returns: a promise that resolves when the request can be issued without queueing,
-  //    or rejects if the request has been cancelled (by the callback)
-  scheduleRequest(handle, callback = () => 0) {
+  // Returns: a promise that resolves to a request token when the request can be issued without queueing,
+  //    or `false` if the request has been cancelled (by getPriority)
+  scheduleRequest(handle, getPriority = () => 0) {
     // Allows throttling to be disabled
     if (!this.props.throttleRequests) {
-      return Promise.resolve(handle);
+      return this._resolveRequest(handle);
     }
 
     // dedupe
-    if (this.requestMap[handle.id]) {
-      return this.requestMap[handle.id];
+    if (this.requestMap.has(handle)) {
+      return this.requestMap.get(handle);
     }
 
-    let request = null;
+    const request = {handle, getPriority};
     const promise = new Promise((resolve, reject) => {
-      request = {handle, callback, resolve, reject};
+      request.resolve = resolve;
+      request.reject = reject;
       return request;
     });
 
-    // TODO - error, request is not defined?
-    // @ts-ignore
-    this.requestQueue.push({promise, ...request});
-    this.requestMap[handle.id] = promise;
+    this.requestQueue.push(request);
+    this.requestMap.set(handle, promise);
     this._issueNewRequests();
     return promise;
   }
 
+  // PRIVATE
+  _resolveRequest(handle, resolve) {
+    let isDone = false;
```

Comment:
The isDone logic is quite hard to understand. Can it be written more clearly or more generously commented.

My add a comment explaining why are we resolving requests many times, and what we are doing about it.

It seems if we call this many times we create different functions and can still end up calling _endRequest mode than once?